app/modules/mqtt.c

```diff
@@ -355,19 +482,81 @@ static void mqtt_socket_received(void *arg, char *pdata, unsigned short len)
       break;
 
     case MQTT_DATA:
-      mud->mqtt_state.message_length_read = length;
-      mud->mqtt_state.message_length = mqtt_get_total_length(in_buffer, mud->mqtt_state.message_length_read);
+      message_length = mqtt_get_total_length(in_buffer, in_buffer_length);
       msg_type = mqtt_get_type(in_buffer);
       msg_qos = mqtt_get_qos(in_buffer);
-      msg_id = mqtt_get_id(in_buffer, mud->mqtt_state.message_length);
+      msg_id = mqtt_get_id(in_buffer, in_buffer_length);
+
+      NODE_DBG("MQTT_DATA: msg length: %u, buffer length: %u\r\n",
+           message_length,
+           in_buffer_length);
+
+      if(message_length > mud->connect_info.max_message_length) {
+        // The pending message length is larger than we was configured to allow
+        NODE_DBG("MQTT: msg too long: total=%u, deliver=%u\r\n", message_length, in_buffer_length);
+        if (msg_type == MQTT_MSG_TYPE_PUBLISH) {
+          // In practice we should never get any other types..
+          deliver_publish(mud, in_buffer, in_buffer_length, 1);
+        }
+
+        if (message_length > in_buffer_length) {
+          // Ignore bytes in subsequent packet(s) too.
+          NODE_DBG("MQTT: skipping into next rx\n");
+          mud->mqtt_state.recv_buffer_state = MQTT_RECV_SKIPPING;
+          mud->mqtt_state.recv_buffer_skip = (uint32_t)message_length - in_buffer_length;
+          break;
+        } else {
+          NODE_DBG("MQTT: Skipping message\n");
+          mud->mqtt_state.recv_buffer_state = MQTT_RECV_NORMAL;
+          goto RX_MESSAGE_PROCESSED;
+        }
+      } else {
+        if (message_length == -1 || message_length > in_buffer_length) {
+          // Partial message in buffer, need to store on heap until next RX. Allocate size for full message directly,
+          // instead of potential reallocs, to avoid fragmentation.
+          // If message_length is indicated as -1, we do not have enough data to determine the length properly.
+          // Just put what we have on heap, and place in state BUFFERING_SHORT.
+          NODE_DBG("MQTT: Partial message received (%u of %d). Buffering\r\n",
+              in_buffer_length,
+              message_length);
+
+          // although message_length is 32bit, it should never go above 16bit since
+          // max_message_length is 16bit.
+          uint16_t alloc_size = message_length > 0 ? (uint16_t)message_length : in_buffer_length;
+
+          mud->mqtt_state.recv_buffer = c_zalloc(alloc_size);
```

Comment:
Can I get here with an existing recv_buffer allocated?

Reply:
Should not be possible (but I may have overseen something of course). If it was in BUFFERING before, and recv_buffer was already allocate, then we should never get down here until  message_length == in_buffer_length (line 361, `in_buffer_length < mud->mqtt_state.recv_buffer_size` check which stops further processing).